### ISSUE TITLE
cut-and-paste-mixin in standard-output-stream

### ISSUE THREAD

**Issue by ormf** [McCLIM/McCLIM#518]
The cut-and-paste-mixin in the standard-output-stream class implements a special behaviour for shift clicking, making it impossible to define custom gestures reacting to shift-click events in an application-pane. Circumventing this in a new application would mean to rename/redefine a custom version of standard-extended-output-stream, clim-stream-pane and application-pane with all its associated methods (if any). Wouldn't it be more reasonable to remove the cut-and-paste mixin from a standard-output-stream? If this is needed for backward compatibility, I'd volunteer to rewrite the code to facilitate for an init-argument :cut-and-paste-mixin which defaults to t but can be disabled in order to be able to use custom shift-click behaviour.

**Comment by dkochmanski** [comment]
I wonder whenever we could define text-selection-protocol in terms of commands and pull them into a command table. In that case later defined commands will overwrite the gesture associated with a translator. I didn't read into spec for that, but that would be my intuition of how we could try tackling this issue.

**Comment by ormf** [comment]
I only partly understand this but it at least sounds better than my solution ;-). I'll go ahead, read docs and code to fully understand what you're saying and will get back...

**Comment by lokedhs** [comment]
I've also looked into text selection and I agree with what has been said above, from both authors. 

I have several issues with the current implementation: first and foremost it's not a standardised interface. Yiu can select text in stream panes but not in Drei panes. In Drei you have to use Emacs-like selection (I think). 

Secondly, using shift and drag is quite unintuitive. 

Thirdly, currently only selection is implemented, and only for text. Most users would like to be able to use the clipboard. 

The current implementation is very much focused on traditional Unix text selection (before the clipboard was commonly used). Thus, three is no actual copy or cut operation and as soon as some text is selected, it goes to the current selection. This behaviour would obviously not be very popular on Windows where the selection is not shared, and you need to perform a copy for it to be made to available to other applications. Overwriting the clipboard every time a selection is made would not be very popular I believe. 

I've played around with this a bit, and my work in progress is part of my Maxima client. It's not ready for production yet, but the general idea is that you register objects as being the current selection or clipboard. The generic function is responsible for converting the object to some format that the requesting client wants. By default, strings are supported, with conversions to the different string formats that exist. Conversion to HTML would also be trivial.

**Comment by ormf** [comment]
Hi Daniel, I think I got your idea and it sounds feasonable. In order to do this it'd be good to have a working example for the text selection protocol to make sure it still works after the change. I couldn't find anything in the clim-examples (at least the text edit fields don't seem to react to shift-click/drag/release actions as I'd expect). Are you aware of a working example? I'd like to distill some minimum example from this for testing.

**Comment by dkochmanski** [comment]
there is no particular protocol to it. If you want to test if it still works, you may try selecting names in address-book demo. Only information about behavior is stored in comment top-level in file text-selection.lisp and on this wiki.

@lokedhs I remember cut-paste worked on climacs, which is based on drei. So I think that thix mixin works on drei under some conditions met by climacs.

**Comment by ormf** [comment]
**@dkochmanski** Maybe I was unclear: Selecting names in the AdressBook even works, if the cut-and-paste-mixin is completely removed from standard-output-stream (I actually can't see any difference in the program's behaviour with or without cut-and-paste-mixin). I might have to check out Climacs or Drei to find a working example of the shift-click drag to be able to check whether it still works after I rewrote the code. If nobody makes use of it the mixin could be removed from standard-output-stream altogether as I can't find any place in the specs where this is mentioned. But this should get double-checked. There must be a reason, why it included it in the first place. To me it appears a bit strange that the mixin gets included in the output-stream class definition but redefines the event-handler for input events (probably that's the "paste" part of "copy-and-paste"...)

**Comment by dkochmanski** [comment]
Try pressing shift and selecting part of text with your mouse button pressed (like you do in "normal" applications) – selected text should be highlighted. That text is available in a X selection buffer (you may paste it with middle click on xterm for instance). text-selection.lisp file mentions other shortcuts which allow pasting X selection from other application as well. this is not for selecting element from a list, but rather selecting text in any kind of output-stream.

**Comment by ormf** [comment]
Sorry, I can't make that work: I tried to select Text in many demos from demodemo. In the Adressbook, I input a lot of text and tried to select any Text in the three panes but nothing highlighted (except for the presentation in the appropriate input context on mouse over, but that disappears as soon as I press the shift key). The help minibuffer at the bottom also doesn't show anything as soon as I press the shift key. I also tried to middle-click paste in the terminal but there is nothing in the clipboard. I must be doing something wrong. Will recheck out the code tomorrow and try again. 

BTW: I get this in my repl: 
Warning, unhandled type "text/plain;charset=utf-8". Sending property NIL to target

I don't know when this appeared but maybe that has to do with it...

**Comment by nsrahmad** [comment]
@ormf you have to keep shift pressed when selecting text with mouse. Here's a demo of how it works for me:
![anim](https://user-images.githubusercontent.com/10996250/43758449-c0d78b78-9a39-11e8-803f-24e8f1d4dc43.gif)

**Comment by ormf** [comment]
Cool video, thanks :-)
It didn't work like that here yesterday, but I reinstalled everything and now it works...

**Comment by ormf** [comment]
I tried to work on the command issue. It is not very complicated to write the routines for text selection, but I have problems completely understanding the mechanism: Our command should be invoked on any pane implementing the cut-and-paste mixin, but I don't know how to use "define-command" in a way that it specializes on that class. As far as I understand the mechanism explained in the clim specs, a command normally should get invoked on presentations (or on the "blank-area" of a sheet), but I don't know how to get the mechanism to just get invoked on the whole area of a pane regardless of what's displayed and how to specify the pane class in the define-class form. In the example I found the body of the define-command extracts the instance of the pane from the application-frame. Can someone help?


Can someone help?
